store/tikv/range_task.go

```diff
@@ -0,0 +1,298 @@
+	return &rangeTaskWorker{
+		name:       s.name,
+		store:      s.store,
+		handler:    s.handler,
+		maxBackoff: s.maxBackoff,
+		taskCh:     taskCh,
+		wg:         wg,
+
+		totalCompletedRegions: &s.completedRegions,
+	}
+}
+
+// CompletedRegions returns how many regions has been sent requests.
+func (s *RangeTaskRunner) CompletedRegions() int32 {
+	return atomic.LoadInt32(&s.completedRegions)
+}
+
+// rangeTaskWorker is used by RangeTaskRunner to process tasks concurrently.
+type rangeTaskWorker struct {
+	name       string
+	store      Storage
+	handler    RangeTaskHandler
+	maxBackoff int
+	taskCh     chan *kv.KeyRange
+	wg         *sync.WaitGroup
+
+	completedRegions int32
+
+	totalCompletedRegions *int32
+}
+
+// run runs the worker. It collects all objects from `w.taskCh` and process them one by one.
+func (w *rangeTaskWorker) run(ctx context.Context, cancel context.CancelFunc) {
+	// Each task might take very short time to finish, so if we update the statistics immediately after sending a
+	// request to a region, the contention of atomic operations might reduce the performance. So it's better to
+	// calculate only the current thread and update it to the RangeTaskRunner.
+	ticker := time.NewTicker(rangeWorkerUpdateStatInterval)
+	defer func() {
+		ticker.Stop()
+		w.updateStat()
+		w.wg.Done()
+	}()
+
+	for r := range w.taskCh {
+		select {
+		case <-ctx.Done():
+			break
+		case <-ticker.C:
+			w.updateStat()
+		default:
+		}
+
+		err := w.runForRange(ctx, r.StartKey, r.EndKey)
+
+		if err != nil {
+			logutil.Logger(ctx).Info("canceling range task because of error",
+				zap.String("name", w.name),
+				zap.Error(err))
+			cancel()
+			return
+		}
+	}
+}
+
+func (w *rangeTaskWorker) updateStat() {
+	atomic.AddInt32(w.totalCompletedRegions, w.completedRegions)
+	w.completedRegions = 0
+}
+
+// runForRange runs task for the given range. The range might contains multiple regions, because after the range sent
+// from the master, the region may split.
+func (w *rangeTaskWorker) runForRange(ctx context.Context, startKey []byte, endKey []byte) error {
+	if len(endKey) != 0 && bytes.Compare(startKey, endKey) >= 0 {
+		return nil
+	}
+
+	key := startKey
+
+	for {
+		bo := NewBackoffer(ctx, w.maxBackoff)
+		loc, err := w.store.GetRegionCache().LocateKey(bo, key)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Compare only once
+		isLast := len(loc.EndKey) == 0 || (len(endKey) > 0 && bytes.Compare(loc.EndKey, endKey) >= 0)
+
+		r := kv.KeyRange{
+			StartKey: key,
+			EndKey:   endKey,
+		}
+		if !isLast {
+			r.EndKey = loc.EndKey
+		}
+
+		nextKey, err := w.handler(ctx, bo, r, loc)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		key = r.EndKey
+		if len(nextKey) > 0 {
+			key = nextKey
+		} else {
+			// If nextKey is set, it means the current region should be processed again. So calculate stats only when
+			// nextKey is not set.
+			w.completedRegions++
+		}
+
+		if isLast {
+			return nil
```

Comment:
when `nextKey` is not nil, it means current region is not completed (even if it is the last region)

Reply:
Fixed at L272